### PR TITLE
spawn(windows): release StaticPipeWriter start() ref when buffer-stdin write completes

### PR DESCRIPTION
**Pull request by robobun**
On Windows, `Bun.spawn`/`Bun.spawnSync` with a non-empty `ArrayBuffer`/`Blob` `stdin` leaked one `StaticPipeWriter` allocation per spawn.

## Cause

`StaticPipeWriter::start()` takes a `+1` and sets `started = true`. #30875 added a `release_start_ref` block to `on_write()` that clears `started` and derefs once the buffer drains, but gated it behind `#[cfg(not(windows))]`:

https://github.com/oven-sh/bun/blob/892b1dabc6/src/spawn/static_pipe_writer.rs#L228-L241

On Windows, when the `uv_write` completes before the child exits (the common ordering for a small buffer):

```
on_write_complete(status=0)
  -> Parent::on_write(Pending)             (release_start_ref gated out)
     -> writer.close()
        -> BaseWindowsPipeWriter::close -> on_close_source
           -> StaticPipeWriter::on_close -> Subprocess::on_close_io(Stdin)
              stdin <- Writable::Ignore; buffer.deref()   (create()'s +1)
  scopeguard deref                                         (write()'s +1)
```

Net: `started` is still `true` but `stdin` is `Writable::Ignore`, so the other release site, `take_pending_start_writer` (called from `on_process_exit` / `close_io`), matches nothing and start()'s `+1` is stranded. The error arm of `on_write_complete` (e.g. `ERROR_BROKEN_PIPE` when the child closes stdin without reading) reaches `on_close` via `close()` without ever calling `Parent::on_write`, so start()'s `+1` is stranded there the same way.

`ShellSubprocess` has the same leak for `cmd < ${buf}` stdin since its `on_process_exit`/`close_io` never look at `started` at all and depend entirely on `on_write`'s release.

## Fix

- `StaticPipeWriter::on_write`: drop the `cfg(not(windows))` gates so `release_start_ref` runs on every platform. `WindowsBufferedWriter` never passes `EndOfFile`, so the existing `status != EndOfFile` guard is a no-op there and is left for the POSIX path (`PosixBufferedWriter::_on_write` still touches `self` after the callback on `EndOfFile`).
- `StaticPipeWriter::on_close` (Windows only): claim `started` and deref after `on_close_io` for paths that reach `close()` without going through `on_write` (the `on_write_complete` error arm). `write()`'s +1, held by that callback's scopeguard, keeps `self` live past the release. This is gated to Windows because on POSIX `drain_buffered_data` may call `on_error()` -> `close()` -> `on_close` and then `on_write()` on the same object with no extra ref held; #34697 changes that ordering and the gate can go once it lands.
- `Subprocess::take_pending_start_writer`: clear `started` when it claims the pointer. The three release sites are then mutually exclusive via `started`, covering the ordering where `on_process_exit` closes the pipe while a `uv_write` whose I/O already completed is still queued (libuv delivers that callback with its real status after `uv_close`, not `ECANCELED`).
- `WindowsBufferedWriter::on_write_complete`: `saturating_sub` for `has_pending_data` so the late-success-after-close ordering does not underflow in debug builds.

`SecurityScanSubprocess` already derefs and writes `started = false` immediately after `start()`, so both new release sites see `started = false`; no change in its refcount balance.

## Verification

Verified on Windows x64 by instrumenting `Drop` with a live counter and spawning 400 children per mode: unfixed leaves 400 live writers after each of the drain (64 B into `sort`) and reject (256 KB into `cmd /c exit`) workloads; fixed returns to 0 on both, in release and debug builds.

Also verified on Windows x64: `spawn.test.ts -t 'stdin|Uint8Array|Blob'` (45/45), `spawnSync.test.ts` (6/6), `file-io.test.ts` (26/26), `child_process.test.ts` (41/41), and on Linux: `bunshell.test.ts` (414/414), `bun-security-scanner-workspaces.test.ts` (3/3). `rust:check-all` is green across all 10 targets.

Found while working on #35286, which scopes its change to the empty-buffer path and does not overlap.

### CODE REVIEW

File: src/spawn/static_pipe_writer.rs [hunk 1]
```diff
@@ -226,14 +226,18 @@ impl<P: StaticPipeWriterProcess> StaticPipeWriter<P> {
...
             let release_start_ref = self.started && status != WriteStatus::EndOfFile;
-            #[cfg(not(windows))]
             if release_start_ref {
                 self.started = false;
             }
             self.writer.close();
-            #[cfg(not(windows))]
             if release_start_ref {
                 // SAFETY: start()'s +1 was still outstanding; `started` cleared above so no other site re-derefs.
                 unsafe { RefCount::<Self>::deref(std::ptr::from_mut::<Self>(self)) };
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/spawn/static_pipe_writer.rs:243]: Addressed in cc2cb051d5 with the `on_close` release gated to `cfg(windows)`, plus a second test case for the error arm (256 KB into `cmd /c exit`: 28 MB unfixed, ~350 KB fixed over 3000 spawns).

I kept it Windows-only rather than unconditional because on POSIX `drain_buffered_data` calls `on_error()` (which reaches `close()` and `on_close` via `_on_error`) and then `on_write()` on the same object with no keepalive held, so releasing both refs in `on_close` there would free the writer before `on_write` runs. On Windows `write()`'s +1 (held by `on_write_complete`'s scopeguard) keeps the writer live past the release. #34697 changes `drain_buffered_data` to stop calling `on_error` then `on_write` in that order; once that lands the gate can be dropped.

File: test/js/bun/spawn/spawn-buffer-stdin-leak-fixture.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+//            the pipe buffer and completes successfully before the child
+//            exits, so `StaticPipeWriter::on_write` runs the close path.
+//   "reject": 256 KB into `cmd /c exit` (never reads stdin). The write
+//            overflows the pipe buffer and the child closes the read end,
+//            so `WindowsBufferedWriter::on_write_complete` takes its error
+//            arm (`close()` then `on_error()` with no `Parent::on_write`).
+
+const mode = process.argv[2] === "reject" ? "reject" : "drain";
+const stdinBuf = mode === "drain" ? Buffer.alloc(64, "x") : Buffer.alloc(256 * 1024, "x");
+const childCmd = mode === "drain" ? ["sort"] : ["cmd", "/c", "exit"];
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/spawn/spawn-buffer-stdin-leak-fixture.ts]: `Buffer`, `ArrayBuffer`, `Uint8Array`, and `Blob` stdin all converge on the same `Writable::Buffer(StaticPipeWriter::create(...))` before `start()` ever runs; the refcount path this PR touches does not branch on input kind, so one representative is enough. `Bun.spawnSync` on Windows drives the same writer through the same `on_write_complete` / `on_close` callbacks on an isolated loop. Adding the full 12-cell matrix at 6000 spawns each would push the file past two minutes under debug for coverage that exercises the same release sites, so I am keeping the two tests that exercise the two distinct release paths.

File: test/js/bun/spawn/spawn-buffer-stdin-leak-fixture.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      procs.push(proc.exited);
+    }
+    await Promise.all(procs);
+  }
+}
+
+function settle() {
+  Bun.gc(true);
+  return Bun.sleep(50).then(() => Bun.gc(true));
+}
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/spawn/spawn-buffer-stdin-leak-fixture.ts]: Added a comment on `settle()` explaining the delay lets the `uv_close` callbacks and pipe handle frees queued by the last batch run before sampling RSS.

File: test/js/bun/spawn/spawn-buffer-stdin-leak.test.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    async () => {
+      const deltaKB = await measure("drain");
+      // Unfixed: one `StaticPipeWriter` (with its embedded `uv_write_t`) per
+      // spawn, ~800 B each, so ~2.4 MB over 3000 spawns in release and more
+      // under debug/ASAN redzones. Fixed: the same workload is within a few
+      // hundred KB of the `stdin: "ignore"` baseline.
+      const boundKB = isASAN || isDebug ? 1200 : 800;
+      expect(deltaKB).toBeLessThan(boundKB);
+    },
+    120_000,
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/spawn/spawn-buffer-stdin-leak.test.ts]: Each case spawns 6000 processes and takes ~14 s under a debug build; the 5 s default kills it mid-warmup. This matches the existing pattern in `spawn-stdout-iterate-leak.test.ts` (300 s) for leak tests that have to iterate.

File: src/runtime/api/bun/subprocess.rs [hunk 5]
```diff
@@ -211,10 +211,22 @@ const _: () = {
 };
 
 impl<'a> Subprocess<'a> {
+    /// Claim `start()`'s outstanding +1 on the buffer-stdin writer (if any)
+    /// for the caller to `deref()`. Clears `started` so `StaticPipeWriter::
+    /// on_write`'s own release site becomes a no-op if a queued write-complete
+    /// callback still fires after the caller closes the pipe (Windows: a
+    /// `uv_write` whose I/O already completed is delivered with its real
+    /// status after `uv_close`, not `ECANCELED`).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/api/bun/subprocess.rs:219]: Included the one-liner in 3e81ba8124: `pending.len().saturating_sub(written)` at PipeWriter.rs:1513. With `on_close` having cleared the buffer and `is_done = true`, the late callback now reports `Drained` and `StaticPipeWriter::on_write` sees `started = false` and `source = None`, so the close branch is a no-op and the scopeguard releases the last ref.

File: test/js/bun/spawn/spawn-buffer-stdin-leak-fixture.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    for (let j = 0; j < BATCH && i + j < count; j++) {
+      const proc = Bun.spawn({
+        cmd: [...childCmd],
+        stdin: stdinBuf,
+        stdout: "ignore",
+        stderr: "ignore",
+      });
+      procs.push(proc.exited);
+    }
+    await Promise.all(procs);
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/spawn/spawn-buffer-stdin-leak-fixture.ts]: Added a per-batch exit-code check so a missing or crashing child surfaces as a thrown error (caught by the `expect(stderr).toBe("")` in the test) rather than a 0 KB delta.

File: test/js/bun/spawn/spawn-buffer-stdin-leak.test.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  }, 120_000);
+
+  test("when the child rejects stdin (on_close release)", async () => {
+    const deltaKB = await measure("reject");
+    // Unfixed: the error arm of `on_write_complete` strands start()'s +1 the
+    // same way; with a 256 KB buffer the leaked write request is larger and
+    // 3000 spawns grew RSS by ~28 MB in release. Fixed: a few hundred KB.
+    const boundKB = isASAN || isDebug ? 2000 : 1500;
+    expect(deltaKB).toBeLessThan(boundKB);
+  }, 120_000);
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/spawn/spawn-buffer-stdin-leak.test.ts]: Switched to `describe.concurrent`; each fixture samples RSS in its own subprocess so there is no cross-contamination, and the two fixtures together peak at ~80 short-lived `sort`/`cmd` children plus two bun processes, which the Windows runners handle.

File: src/spawn/static_pipe_writer.rs [hunk 8]
```diff
@@ -296,6 +315,7 @@ impl<P: StaticPipeWriterProcess> StaticPipeWriter<P> {
 /// The heap free is handled by `IntrusiveRc` after `drop` returns.
 impl<P: StaticPipeWriterProcess> Drop for StaticPipeWriter<P> {
     fn drop(&mut self) {
+        eprintln!("SPW_DROP");
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/spawn/static_pipe_writer.rs]: Replaced in ddf245e1 with `LIVE_COUNT.fetch_sub(1, Ordering::Relaxed)` and a `staticPipeWriterLiveCount()` hook in `bun:internal-for-testing`; the test now asserts on the live count instead of RSS.

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Delete the test and instrumentation code

File: src/spawn/static_pipe_writer.rs [hunk 9]
```diff
@@ -264,13 +268,28 @@ impl<P: StaticPipeWriterProcess> StaticPipeWriter<P> {
...
         self.source.detach();
         // SAFETY: `process` is a backref to the owning process, guaranteed alive
         // for the lifetime of this writer (the process owns/outlives its stdio writers).
         unsafe { P::on_close_io(self.process, StdioKind::Stdin) };
+        #[cfg(windows)]
+        if release_start_ref {
+            // SAFETY: `started` was the token for start()'s outstanding +1;
+            // cleared above so no other site re-derefs. Last use of `self`.
+            unsafe { RefCount::<Self>::deref(std::ptr::from_mut::<Self>(self)) };
+        }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/spawn/static_pipe_writer.rs:292]: The test, fixture, `LIVE_COUNT` atomic, and `staticPipeWriterLiveCount()` hook were removed at Jarred's request (https://github.com/oven-sh/bun/pull/35297#issuecomment-5063779003: "Delete the test and instrumentation code"). The verification numbers in the PR body were captured with that instrumentation before removal. Leaving the diff as the fix only per that instruction.
